### PR TITLE
Narrator Changes

### PR DESCRIPTION
**Pull request by CalebKendra**
Changing Narrator to accept a dictionary of values instead of a float value in order to better differentiate between `acts` and `scenes`

### CODE REVIEW

File: libs/narrator/narrator/Path.py [hunk 1]
```diff
@@ -2,15 +2,13 @@
...
 
   def next_scene(self):
     self.scene += 1
 
-  def change(self, outcome: float):
-    outcome = str(float(outcome))
-    path, scene = outcome.split(".")
-    self.number = int(path)
-    self.scene = int(scene)
+  def change(self, path):
```
Comment (CalebKendra) [inline libs/narrator/narrator/Path.py:12]: @dluman what should the default value be for `path` in this function?

Comment (dluman) [reply libs/narrator/narrator/Path.py:12]: Pass a `dictionary` in instead with default values of `1`, `1`?

Comment (dluman) [reply libs/narrator/narrator/Path.py:12]: I was working on this, and I see the issue. I'll `push` something in a bit.

Comment (CalebKendra) [reply libs/narrator/narrator/Path.py:12]: Ah yes that makes sense, ill change that

File: libs/narrator/narrator/Path.py [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
-
 class Path:
 
-  def __init__(self, path: int = 0):
-    self.number = path
-    self.scene = 0
+  def __init__(self, paths: dict = {}):
+    """ Constructor """
+    self.act = paths["act"]
+    self.scene = list(list(paths["scene"])[0].keys())[0]
```
Comment (dluman) [inline libs/narrator/narrator/Path.py:6]: Not enamored with what I did here, but it works.

File: libs/narrator/narrator/Path.py [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
 
-  def next_scene(self):
-    self.scene += 1
+  def get_next_scene(self, paths):
+    scenes = list(
+        paths[self.act]
+    )
+    idx = scenes.index(self.scene)
+    if(idx + 1 <= len(scenes)):
+        self.scene = paths[self.act][scenes[idx+1]]
```
Comment (dluman) [inline libs/narrator/narrator/Path.py:14]: If there's another scene to move to, move to it sequentially.

File: libs/narrator/narrator/Path.py [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
 
-  def change(self, outcome: float):
-    outcome = str(float(outcome))
-    path, scene = outcome.split(".")
-    self.number = int(path)
-    self.scene = int(scene)
+  def change(self, path: dict = {"act": 1, "scene": 1}):
+    """ Change a scene in a currently-running path """
+    self.act = path["act"]
+    self.scene = path["scene"]
```
Comment (dluman) [inline libs/narrator/narrator/Path.py:19]: This seems to work?

File: libs/narrator/narrator/Path.py (hunk 3 above)
Comment (CalebKendra) [reply libs/narrator/narrator/Path.py:14]: ah sorry I forgot to fix that

File: libs/narrator/narrator/Narrator.py [hunk 5]
```diff
@@ -8,17 +8,46 @@ class Narrator:
...
+          for idx in range(scenes):
+            lines += chosen_path.values()[idx]
     else:
-      lines = chosen_path[self.path.scene]
+      # Play one
+      try:
+        lines = chosen_path[self.path.scene]
+      except KeyError:
+        scenes = list(chosen_path)
+        lines = chosen_path[scenes[0]]
```
Comment (dluman) [inline libs/narrator/narrator/Narrator.py:47]: This is the partial scene play structure I referred to in my Discord message.

File: libs/narrator/narrator/Narrator.py [hunk 6]
```diff
@@ -24,20 +24,24 @@ def narrate(self, **kwargs):
...
+      except ValueError:
+        print("ERROR: Scene value could not be converted to an integer!")
+        exit()
+      try:
+        scene_location = list(chosen_path.keys()).index(self.path.scene)
+        for idx in range(scene_location,scene_location+scenes):
+          lines += list(chosen_path.values())[idx]
+      except IndexError:
+        print("ERROR: Scene value is greater then remaining amount of scenes!")
+        exit()
```
Comment (CalebKendra) [inline libs/narrator/narrator/Narrator.py:44]: When you put in the "scene" you wanted to start at it did not actually start there, so I changed it so its starts at the given scene

**Review by dluman — APPROVED** [review]
Good pass to make this more flexible. We need to remember to reformat the old assignments that use `narrator`.

**Pull request merged**